pkg/pgmodel/querier/query_builder.go

```diff
@@ -379,13 +411,55 @@ type aggregators struct {
 	timeParams  []interface{}
 	valueClause string
 	valueParams []interface{}
+	unOrdered   bool
+}
+
+/* vectorSelectors called by the timestamp function have special handling see engine.go */
+func calledByTimestamp(path []parser.Node) bool {
+	if len(path) > 0 {
+		node := path[len(path)-1]
+		call, ok := node.(*parser.Call)
+		if ok && call.Func.Name == "timestamp" {
+			return true
+		}
+	}
+	return false
 }
 
+var vectorSelectorExtensionRange = semver.MustParseRange(">= 0.1.3-beta")
+
 /* The path is the list of ancestors (direct parent last) returned node is the most-ancestral node processed by the pushdown */
-func getAggregators(hints *storage.SelectHints, path []parser.Node) (*aggregators, parser.Node, error) {
-	if extension.ExtensionIsInstalled && path != nil && hints != nil && len(path) >= 2 && !hasSubquery(path) {
+func getAggregators(hints *storage.SelectHints, qh *QueryHints, path []parser.Node) (*aggregators, parser.Node, error) {
+	if extension.ExtensionIsInstalled && qh != nil {
+		//switch on the current node being processed
+		switch n := qh.CurrentNode.(type) {
+		case *parser.VectorSelector:
+			//TODO: handle the instant query (hints.Step==0) case too.
+			if hints != nil &&
+				hints.Step > 0 &&
+				hints.Range == 0 &&
+				!calledByTimestamp(path) &&
+				n.OriginalOffset == time.Duration(0) &&
+				n.Offset == time.Duration(0) &&
+				!hasSubquery(path) &&
+				vectorSelectorExtensionRange(extension.PromscaleExtensionVersion) {
+				qf := aggregators{
+					timeClause:  "ARRAY(SELECT generate_series($%d::timestamptz, $%d::timestamptz, $%d))",
```

Comment:
It is difficult for me to understand how it helps. Can you please write a brief summary so that it's easy to follow?